lib/datadog/core/workers/async.rb

```diff
@@ -17,9 +17,24 @@ def self.included(base)
 
           # Methods that must be prepended
           module PrependedMethods
+            def initialize(*args, **kwargs, &block)
```

Comment:
Is there a reason this needs to be done in the constructor?

Generally speaking its unsafe to prepend onto `initialize`, because not all objects have the same constructor signature. It introduces a kind of unspoken coupling (expects implementing objects to have a very specific constructor.)

Reply:
Attempting to answer my own question, I'm going to guess this has something to do with lazy initialization not being sufficient if that object is accessed concurrently prior to initialization...

Reply:
FWIW the breakage I'm seeing in CI right now is related to the constructor definition:

```
68.2) Failure/Error:
              def initialize(&block)
                @task = block
              end
            
            ArgumentError:
              wrong number of arguments (given 1, expected 0)
            # ./lib/datadog/core/worker.rb:8:in `initialize'
```

But it's only on Ruby 2.1-2.4. Probably the `kwargs`, so we may need to introduce some compatibility checks there.

Reply:
> Attempting to answer my own question, I'm going to guess this has something to do with lazy initialization not being sufficient if that field is accessed concurrently prior to initialization...

Yeah, there's a chance that one thread will use one mutex instance, and then another will use another one. If we can move mutex into a constant — that would solve the issue entirely :D but can cause an issue with performance (haven't looked into details how those classes are used).